### PR TITLE
Rework pronoun extraction

### LINKED ISSUE #39: new field matcher matches less than before
Note: matches less than before in some instances, i.E. partial fieldname matches (ex: "pronouns EN")

_Originally posted by @se4598 in https://github.com/ItsVipra/ProToots/pull/38#discussion_r1252328545_
            
whoopsie daisy i messed up.
what if we just use two nested loops instead?
makes the runtime n^2 but uuuhhh

### PR DESCRIPTION
**Pull request by nachtjasmin**
Closes #39, #35, #28, #7.

### CODE REVIEW

File: src/libs/pronouns.js [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+function sanitizePronounPageValue(val) {
+	if (!val.startsWith("https://")) return val;
+
+	val = val.replace(/https?:\/\/.+\.pronouns\.page\/:?/, "");
+
+	if (val === "no-pronouns") val = "no pronouns";
+	return val;
+}
+
+const pronounAllowlist = [
```
Comment (se4598) [inline src/libs/pronouns.js:123]: pronounsAllowlist sounds wrong (as in: to allow/deny pronouns).
something with recognizable, common or known? suggestions: "knownPronounsList"

Comment (nachtjasmin) [reply src/libs/pronouns.js:123]: You're right, I should change this. Thanks!

File: src/libs/pronouns.js [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+		if (knownPronouns.includes(subjective) && knownPronouns.includes(objective)) {
+			return match.replaceAll(" ", "");
+		}
+	}
+
+	const followedByColon = bio.matchAll(/pronouns?:\W+([\w/+]+)/gi);
+	for (const match of followedByColon) {
+		return match.pop(); // first group is last entry in array
+	}
+
```
Comment (ItsVipra) [inline src/libs/pronouns.js:230]: it looks like we're missing out on "any/all" or "any pronouns".

how about another matcher like `/[any|all]+ pronouns/gi`?

Comment (nachtjasmin) [reply src/libs/pronouns.js:234]: Added those.

**Comment by nachtjasmin** [comment]
Just realised that we can also close #33 with it. We don't have HTML in our pronouns anymore, so we could also get rid of setting the `innerHTML`.

gonna provide a commit with that soon.

**Comment by nachtjasmin** [comment]
Also, [this method](https://github.com/ItsVipra/ProToots/blob/f51b2e71fe6e804cf9ac3a1a4096397b0bd2d640/src/libs/protootshelpers.js#L28) should be moved to pronouns.js.

File: src/libs/pronouns.js [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+		if (lang in profiles) {
+			pronouns = profiles[lang].pronouns;
+			break;
+		}
+	}
+
+	// If we don't have a value yet, just take the first profile.
+	if (!pronouns) pronouns = profiles[0].pronouns;
+
+	let val = pronouns.find((x) => x.opinion === "yes").value;
```
Comment (ItsVipra) [inline src/libs/pronouns.js:107]: the api labels thumbs up/okay as "meh", i think "okay" pronouns should still count
```suggestion
	let val = pronouns.find((x) => x.opinion === "yes" || x.opinion === "meh").value;
```

**Comment by ItsVipra** [comment]
> Just realised that we can also close #33 with it. We don't have HTML in our pronouns anymore, so we could also get rid of setting the `innerHTML`.
> 
> gonna provide a commit with that soon.

but the plate itself is still in the a tag - which was the original issue you had, right?
image

File: src/libs/pronouns.js [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+	let text = sanitizeHtml(pronounsRaw, { allowedTags: [], allowedAttributes: {} });
+	// If one of pronoun URLs matches, overwrite the current known value.
+	for (const knownUrlRe of knownPronounUrls) {
+		if (!knownUrlRe.test(pronounsRaw)) continue;
+		text = pronounsRaw.match(knownUrlRe)[1];
+	}
+
+	// Right now, only the pronoun.page regex matches the @usernames.
+	if (text.charAt(0) === "@") {
+		text = await queryPronounsFromPronounsPage(text.substring(1));
```
[the remark this answers is not in the document]
Comment (nachtjasmin) [reply src/libs/pronouns.js:66]: We don't need to, no. However, it's hard to distinguish if we already have a valid pronoun field. Like, for all the cases below, how should we differentiate/handle them?

- `she/her https://en.pronouns.page/@account`
- `:they: or something else :extremely_gay: :european_flag: :whatever: https://en.pronouns.page/@account`
- `de: sie; en: she; gender: not existing; https://de.pronouns.page/@account`
- `idk what my pronouns are (still here's a link https://en.pronouns.page/@account)`

I've decided to go the easy way and fetch the API, even if it's not necessary at all. That way I can be sure that we display pronouns and not some other fuckery. ^^

[the remark this answers is not in the document]
Comment (ItsVipra) [reply src/libs/pronouns.js:66]: fair enough - easier than trying to parse natural language '^^

**Pull request merged**